.circleci/integration-tests/airflow_dag_introspection.py

```diff
@@ -0,0 +1,174 @@
+        print(f"Found logs: {logs}")
+        assert notexpect not in logs  # nosec: B101
+        assert expect in logs  # nosec: B101
+        print(f"Found {expect} but not {notexpect}")
+
+    # make sure expected output appeared
+    check(this_task_instance, expected, notexpected)
+
+
+def assert_homomorphic(task_group_names, **context):
+    """The structure of all of the task groups above should be the same"""
+    # get the dag in dot notation, focus only on its edges
+    dag = context["dag"]
+    print(dag)
+    # gives string which represents whole dag structure
+    graph = render_dag(dag)
+    print("Whole DAG:")
+    print(indent(str(graph), "    "))
+    lines = list(filter(lambda x: "->" in x, str(graph).split("\n")))
+
+    # bin them by task group, then remove the group names
+    group_strings = []
+    # removes everything that's not a task name
+    for name in task_group_names:
+        print(name)
+        relevant_lines = filter(lambda x: name in x, lines)
+        normalized_lines = (x.strip().replace(name, "") for x in sorted(relevant_lines))
+        edges_str = "\n".join(normalized_lines)
+        group_strings.append(edges_str)
+        print(indent(edges_str, "    "))
+
+    # these should be identical
+    for xgroup, ygroup in zip(group_strings, group_strings[1:]):
+        assert xgroup == ygroup  # nosec: B101
+
+
+def get_the_task_states(task_ids: list[str, str, str], **context) -> dict[str, str]:
+    """This function returns a dictionary of task_ids and a tasks state from a list of task_ids"""
+    dag_instance = context["dag"]
+    logical_date = context["logical_date"]
+
+    ls_of_statuses = []
+    for i in task_ids:
+        j = TaskInstance(dag_instance.get_task(i), execution_date=logical_date).current_state()
+        ls_of_statuses.append(j)
+
+    for i, j in zip(task_ids, ls_of_statuses):
+        print(f"The state for the task with task_id: '{i}' is state: '{j}'")
+
+    dict_of_results = {task_ids[i]: ls_of_statuses[i] for i in range(len(task_ids))}
+    return dict_of_results
+
+
+def assert_the_task_states(task_ids_and_assertions: dict[str, str], **context):
+    """This function makes an assertion that supposed task states are actually that state"""
+    """If the tasks are that state then it returns the value passed in, unaltered."""
+    dag_instance = context["dag"]
+    logical_date = context["logical_date"]
+
+    ls_of_statuses = []
+    for task_id in task_ids_and_assertions.keys():
+        current_state = TaskInstance(
+            dag_instance.get_task(task_id), execution_date=logical_date
+        ).current_state()
+        ls_of_statuses.append(current_state)
+
+    for task_id, state, actual_state in zip(
+        task_ids_and_assertions.keys(), task_ids_and_assertions.values(), ls_of_statuses
+    ):
+        print(f"The state for the task with task_id: '{task_id}' is state: '{actual_state}'")
+        assert state == actual_state  # nosec: B101
```

Comment:
After a second thought, as we're actually running these lines on astro cloud, should we raise another exception when these conditions aren't met instead of using assert?